pysph/tools/interpolator.py

```diff
@@ -27,12 +30,113 @@ def post_loop(self, d_idx, d_prop, d_number_density):
             d_prop[d_idx] /= d_number_density[d_idx]
 
 
-class InterpolateSPH(Equation):
-    def initialize(self, d_idx, d_prop):
-        d_prop[d_idx] = 0.0
+class SPHFirstOrderApproximationPreStep(Equation):
+    def __init__(self, dest, sources, dim=1):
+        self.dim = dim
+
+        super(SPHFirstOrderApproximationPreStep, self).__init__(dest, sources)
+
+    def initialize(self, d_idx, d_mommat):
+        i, j = declare('int', 2)
+
+        for i in range(4):
+            for j in range(4):
+                d_mommat[16*d_idx + j+4*i] = 0.0
+
+    def loop(self, d_idx, d_h, s_h, s_x, s_y, s_z, d_x, d_y, d_z, s_rho,
+             s_m, WIJ, XIJ, DWIJ, d_mommat):
+        # Vj = s_m[s_idx] / s_rho[s_idx]
+        Vj = 1.0
+
+        d_mommat[16*d_idx+0] += WIJ * Vj
```

Comment:
You could pre-compute 16*d_idx into a variable and save 16 multiplications.

Reply:
done